src/components/datagrid/column_actions.tsx

```diff
@@ -0,0 +1,195 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { EuiDataGridColumn, EuiDataGridSorting } from './data_grid_types';
+import { EuiI18n } from '../i18n';
+import { EuiListGroupItemProps } from '../list_group';
+import React from 'react';
+
+export function getColumnActions(
+  column: EuiDataGridColumn,
+  columns: EuiDataGridColumn[],
+  setVisibleColumns: (columnId: string[]) => void,
+  setIsPopoverOpen: (value: boolean) => void,
+  sorting: EuiDataGridSorting | undefined,
+  switchColumnPos: (colFromId: string, colToId: string) => void
+) {
+  if (column.actions === false) {
+    return [];
+  }
+  const colIdx = columns.findIndex(col => col.id === column.id);
+  const sortingIdx = sorting
+    ? sorting.columns.findIndex(col => col.id === column.id)
+    : -1;
+  const sortBy = (direction: 'asc' | 'desc' = 'asc') => {
+    if (!sorting) {
+      return;
+    }
+
+    const newSorting =
+      sortingIdx >= 0
+        ? //replace existing entry
+          Object.values({
+            ...sorting.columns,
+            [sortingIdx]: {
+              id: column.id,
+              direction: direction,
+            },
+          })
+        : //append entry
+          [
+            ...sorting.columns,
+            {
+              id: column.id,
+              direction: direction,
+            },
+          ];
+
+    sorting.onSort(newSorting as EuiDataGridSorting['columns']);
+  };
+  const onClickHideColumn = () =>
+    setVisibleColumns(
+      columns.filter(col => col.id !== column.id).map(col => col.id)
+    );
+
+  const onClickSortAsc = () => {
+    setIsPopoverOpen(false);
+    sortBy('asc');
+  };
+
+  const onClickSortDesc = () => {
+    setIsPopoverOpen(false);
+    sortBy('desc');
+  };
+
+  const onClickMoveLeft = () => {
+    setIsPopoverOpen(false);
+    const targetCol = columns[colIdx - 1];
+    if (targetCol) {
+      switchColumnPos(column.id, targetCol.id);
+    }
+  };
+
+  const onClickMoveRight = () => {
+    setIsPopoverOpen(false);
+    const targetCol = columns[colIdx + 1];
+    if (targetCol) {
+      switchColumnPos(column.id, targetCol.id);
+    }
+  };
+
+  const result = [];
+  if (column.actions?.showHide !== false) {
+    const option = {
+      label: (
+        <EuiI18n token="euiColumnActions.hideColumn" default="Hide column" />
+      ),
+      onClick: onClickHideColumn,
+      iconType: 'eyeClosed',
+      size: 'xs',
+      color: 'text',
+    };
+    if (typeof column.actions?.showHide === 'object') {
+      result.push({ ...option, ...column.actions?.showHide });
```

Comment:
The optional chaining isn't required on this line, TS [surprisingly, to me!] is smart enough to understand the previous line's implication.

```suggestion
      result.push({ ...option, ...column.actions.showHide });
```